### PR TITLE
Installer key fingerprint

### PR DESCRIPTION
**Pull request by jvatic**
#1298

### CODE REVIEW

File: pkg/awsutil/awsutil.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+func FingerprintPublicKey(key []byte) (string, error) {
+	sshPublicKey, _, _, _, err := ssh.ParseAuthorizedKey(key)
+	if err != nil {
+		return "", err
+	}
+	v := reflect.Indirect(reflect.ValueOf(sshPublicKey))
+	publicKey := &rsa.PublicKey{
+		N: v.FieldByName("N").Interface().(*big.Int),
+		E: v.FieldByName("E").Interface().(int),
+	}
```
Comment (titanous) [inline pkg/awsutil/awsutil.go]: Reaching into unexported structs is a no-no. Instead, you'll have to parse it yourself using (simplified) logic based on the internals of `ssh.ParseAuthorizedKey`.

Comment (jvatic) [inline pkg/awsutil/awsutil.go]: Actually, seeing as we always have the private key for this use case, we can just use that to get the public key in the form of an `rsa.PublicKey`.

File: installer/installer.go [hunk 2]
```diff
@@ -269,19 +270,24 @@ func (s *Stack) createKeyPair() error {
...
+				Filters: []ec2.Filter{{
+					Name:   aws.String("key-name"),
+					Values: []string{keypairName},
+				},
+					{
+						Name:   aws.String("fingerprint"),
+						Values: []string{fingerprint},
+					}},
+			})
+			if err == nil && len(res.KeyPairs) > 0 && *res.KeyPairs[0].KeyName == keypairName {
```
Comment (titanous) [inline installer/installer.go]: I don't really like these nested `err == nil` checks. Is there a way to invert the logic or split it into separate functions?

File: installer/installer.go [hunk 3]
```diff
@@ -269,19 +270,24 @@ func (s *Stack) createKeyPair() error {
...
+			s.SSHKeyName = keypairName
+			res, err := s.ec2.DescribeKeyPairs(&ec2.DescribeKeyPairsRequest{
+				Filters: []ec2.Filter{{
+					Name:   aws.String("key-name"),
+					Values: []string{keypairName},
+				},
+					{
+						Name:   aws.String("fingerprint"),
+						Values: []string{fingerprint},
+					}},
```
Comment (titanous) [inline installer/installer.go]: Split these braces onto two lines (and the same for the opening)

File: installer/installer.go [hunk 4]
```diff
@@ -269,19 +270,24 @@ func (s *Stack) createKeyPair() error {
...
+			})
+			if err == nil && len(res.KeyPairs) > 0 && *res.KeyPairs[0].KeyName == keypairName {
+				// key exists, we're good to go
```
Comment (titanous) [inline installer/installer.go]: This comment is superfluous

**Comment by jvatic** [comment]
@titanous comments addressed

File: installer/installer.go [hunk 5]
```diff
@@ -263,26 +264,46 @@ func (s *Stack) allocateDomain() error {
...
+	}
+	fingerprint, err := awsutil.FingerprintImportedKey(keypair.PrivateKey)
+	if err != nil {
+		return err
+	}
+	res, err := s.ec2.DescribeKeyPairs(&ec2.DescribeKeyPairsRequest{
+		Filters: []ec2.Filter{
+			{
+				Name:   aws.String("key-name"),
+				Values: []string{name},
```
Comment (titanous) [inline installer/installer.go]: Is the name filter necessary?

File: installer/installer.go [hunk 6]
```diff
@@ -263,26 +264,46 @@ func (s *Stack) allocateDomain() error {
...
+			{
+				Name:   aws.String("fingerprint"),
+				Values: []string{fingerprint},
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+	if len(res.KeyPairs) <= 0 {
```
Comment (titanous) [inline installer/installer.go]: `==`

**Comment by jvatic** [comment]
@titanous comments addressed

File: installer/installer.go [hunk 7]
```diff
@@ -263,26 +264,42 @@ func (s *Stack) allocateDomain() error {
...
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+	if len(res.KeyPairs) == 0 {
+		return errors.New("No matching key found")
+	}
+	s.SSHKey = keypair
```
Comment (jvatic) [inline installer/installer.go]: This should use the name from the response as there could be a matching key of another name.

Comment (jvatic) [inline installer/installer.go]: Done.

**Comment by titanous** [comment]
LGTM

**Pull request merged**